### PR TITLE
hychiu02-proposal-submission

### PR DESCRIPTION
**Pull request by hychiu02**
---
A good PR should include the following parts:

- [x] A clear title
- [x] A meaningful message for PR, as well as its commits
- [x] from your specific branch (***not master***) merging to teacher's `HW` branch (***not master***, again)
- [x] excluding any irrelevant files, such as binaries, text files, or dot files
- [ ] passing tests/CI

### CODE REVIEW

File: project/hychiu02/README.md [hunk 1]
```diff
@@ -0,0 +1,96 @@
+# AutoEncoder for arrhythmia detection
+
+## Basic information
+
+[GitHub](https://github.com/hychiu02/AutoEncoder-for-arrhythmia-detection)
+
+Arrhythmia dectection using ECG signal
+
+## Problem to solve
+Heart diseases are one of the significant reasons for death all over the planet. To perform arrhythmia detection, Electrocardiograms(ECGs) is used to monitor hte functioning of heart by capturing electrical activity. Because of needs of real-time inference, the model sizeand  parameters should be small, autoencoder seems to be a good choice. I want to develop a method to build, train and inference a MLP model with C++.
```
Comment (yungyuc) [inline project/hychiu02/README.md:10]: Having a Python wrapper to a small library is a good way to engineer the small library.  Because the library and problem size are small, it should not be used to carry out large amount of computation.  But the (software) engineering has value.

File: project/hychiu02/README.md [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+[GitHub](https://github.com/hychiu02/AutoEncoder-for-arrhythmia-detection)
+
+Arrhythmia dectection using ECG signal
+
+## Problem to solve
+Heart diseases are one of the significant reasons for death all over the planet. To perform arrhythmia detection, Electrocardiograms(ECGs) is used to monitor hte functioning of heart by capturing electrical activity. Because of needs of real-time inference, the model sizeand  parameters should be small, autoencoder seems to be a good choice. I want to develop a method to build, train and inference a MLP model with C++. 
+
+## Prospective Users
+People who want to build Multilayer Perceptron model for particular case.
```
Comment (yungyuc) [inline project/hychiu02/README.md:13]: You can be more specific about the particular cases.  One of the use is the testing and software engineering I mentioned above.

File: project/hychiu02/README.md [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+## System Architecture
+There will be a Node class for storing weight, a Layer class with Nodes for forward and backward weights calculation and a MLP class containing Layers  
+The flow will look like this
```
Comment (yungyuc) [inline project/hychiu02/README.md:17]: I assume the data I/O is handled by using Python?

File: project/hychiu02/README.md [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+![](Flow.png)
+
+
+## API Description
+- Dataset class
+```
+ Dataset
+  the constructor initialize Dataset
+  - parameters:
+    dataPath: ECG file path
```
Comment (yungyuc) [inline project/hychiu02/README.md:27]: Do you require all data to be loaded to memory?

**Review by yungyuc — CHANGES_REQUESTED** [review]
It looks like a complete description for what you want to do and I do not see high risks.  Depending on how you implement it, it may or may not have room for future extension (after the course).

Please address all points I made by replying each of them.  If you agree, simply type ok or agree.  If it's about to change the proposal, your comment should highlight the corresponding change you make.  If you do not agree, we can discuss with your reply.

Please add your project to the list of project, as I have called for your attention in the project page.  For some reasons many of you forget to do it.  If you know others who have not submit the proposal, please remind them.

File: project/hychiu02/README.md (hunk 1 above)
Comment (hychiu02) [reply project/hychiu02/README.md:10]: Maybe I misunderstand the definition of large amount computations.  Initially my thought was that Deep neural network has lots of MACs on forward pass and lots of partial derivative calculation on backward pass. And runing such model on python runtime may not be as efficient as running on C++ runtime. So I implement this method by c++ code and using it on a case. If it works fine on this case, it can be used to other different cases.

File: project/hychiu02/README.md (hunk 2 above)
Comment (hychiu02) [reply project/hychiu02/README.md:13]: Perhaps I misunderstand the meaning of prospective users, I think people who want to do abnormal sound detection, cardiac diseases classification or other regression or classification problem can take advantage of my implementation.

File: project/hychiu02/README.md (hunk 3 above)
Comment (hychiu02) [reply project/hychiu02/README.md:17]: Actually I plan to do only unit test on python. Is it ok for term project?

File: project/hychiu02/README.md (hunk 4 above)
Comment (hychiu02) [reply project/hychiu02/README.md:27]: Actually I plan to load all file names to memory, and when model need data, a part of file will be load.

File: project/hychiu02/README.md (hunk 1 above)
Comment (yungyuc) [reply project/hychiu02/README.md:10]: > Maybe I misunderstand the definition of large amount computations. 

For large amount of computation, we usually refer to more than several cpu-**day**.  It is not uncommon to measure by thousands of cpu-day.  Anything that can be finished in a couple of hours on a single core is not using a large amount of computation.

But the course focuses on the software engineering for the large amount of computation, not the amount of computation itself.  You do not need to find a problem that requires the large amount of computation.  It is good enough that the problem may be extended to use a large amount of computation.  The large amount of computation needs to be justified if it really happens.

> Initially my thought was that Deep neural network has lots of MACs on forward pass and lots of partial derivative calculation on backward pass. And runing such model on python runtime may not be as efficient as running on C++ runtime. 

This is correct.  But IIUC you are doing inferencing (correct me if I am wrong).  Inferential computation is much less than training.

> So I implement this method by c++ code and using it on a case. If it works fine on this case, it can be used to other different cases. Is it kind of large amount computation?

It is not, but it does not invalidate the problem.  It is fine to do it.

File: project/hychiu02/README.md (hunk 2 above)
Comment (yungyuc) [reply project/hychiu02/README.md:13]: You did not misunderstand.  Yes, and please put that in your proposal.  That will make the development process clearer.  And you know what you change when you need to change.

File: project/hychiu02/README.md (hunk 3 above)
Comment (yungyuc) [reply project/hychiu02/README.md:17]: It is OK.  But you did not answer my question.  Please think about how you will do I/O, which will be a hurdle.

File: project/hychiu02/README.md (hunk 4 above)
Comment (yungyuc) [reply project/hychiu02/README.md:27]: Loading data on the fly takes quite some code development and it is hard to have a robust implementation on different platform.  Imagine someone runs your code on NFS.

It's OK.  Just to keep it in mind that I/O is hard.

**Comment by yungyuc** [comment]
Leave a comment if it's ready to be reviewed again.

File: project/hychiu02/README.md (hunk 2 above)
Comment (hychiu02) [reply project/hychiu02/README.md]: Ok, I add into my proposal.Thank you!

File: project/hychiu02/README.md (hunk 3 above)
Comment (hychiu02) [reply project/hychiu02/README.md:17]: I just have raw idea to read this kind of text data by fopen or something. I'll try to survey how to hide I/O survey

File: project/hychiu02/README.md (hunk 4 above)
Comment (hychiu02) [reply project/hychiu02/README.md:27]: Sorry i didn't think well. I'll try to survey how to run across different platform. Currently I just have rough idea to run on my device.

File: project/hychiu02/README.md (hunk 1 above)
Comment (hychiu02) [reply project/hychiu02/README.md:10]: Yeah, you're right. I just want to implemant a simple Machine Learning framework. As my example to predict VA, after training phase with backpropagation, when inferencing with only forward pass, its computation is much less than training. 
Thank you for advising! I'll try to think about the folllowing problem while I'm coding. I add my project to the list of project. Thanks for your kindly reminder.

File: project/hychiu02/README.md (hunk 4 above)
Comment (yungyuc) [reply project/hychiu02/README.md:27]: Blind spots are not something needing to apologize for.  Please do not say sorry for that or it will be hard for the discussions to go productively.

**Comment by hychiu02** [comment]
> Leave a comment if it's ready to be reviewed again.

File: project/hychiu02/README.md (hunk 3 above)
Comment (yungyuc) [reply project/hychiu02/README.md:17]: You may also list it as a limitation of the system and mark is as unhandled in the proposal.

**Review by yungyuc — CHANGES_REQUESTED** [review]
Please resolve the merge conflict before requesting for review again.

File: project/hychiu02/README.md (hunk 3 above)
Comment (hychiu02) [reply project/hychiu02/README.md:18]: Ok, I added it to my proposal. Thank you!

**Review by yungyuc — APPROVED** [review]

**Comment by yungyuc** [comment]
Please let me know when you want it to be merged.

**Comment by hychiu02** [comment]
I think the proposal can be merged, thanks.
